src/os/bluestore/BlueStore.h

```diff
@@ -1826,8 +1828,97 @@ class BlueStore : public ObjectStore,
       trace_period_mcs = rate > 0 ? floor((1/rate) * 1000000.0) : 0;
 #endif
     }
+    int64_t get_kv_throttle_current() {
+      return throttle_bytes.get_current();
+    }
+    void reset_kv_throttle_max(int64_t m) {
+      throttle_bytes.reset_max(m);
+    }
   } throttle;
 
+  /**
+   * CoDel algorithm to calculate the incoming batch size to kv by using Throttle mechanism
+   */
+  class BlueStoreSlowFastCoDel {
```

Comment:
I don't think this needs to be embedded directly into BlueStore.  There seem to be two touchpoints between BlueStoreSlowFastCoDel and BlueStore's internal data structures.

1) submit_txc_info uses a TransContext to get two pieces of information: txc_state_proc_start and bytes.  Explicitely pass in bytes and operation latency instead, as well as the value of bs_throttle->get_kv_throttle_current().
2) bs_throttle only seems to be in one other place: reset_kv_throttle_max.  BlueStoreSlowFastCoDel could simply take a std::function<void(uint64_t)> argument in the constructor.

With those changes, you should be able to move this whole structure into its own .h and .cc files without any references to bluestore internals.  This would also enable you to write a unit test for it independently of bluestore.